provisioner/terraform/resources.go

```diff
@@ -748,6 +750,30 @@ func convertAddressToLabel(address string) string {
 	return cut
 }
 
+func dependsOnAgent(graph *gographviz.Graph, agent *proto.Agent, appAgentID string, resource *tfjson.StateResource) bool {
+	// Plan: we need to find if there is edge between the agent and the resource.
+	if agent.Id == "" && appAgentID == "" {
+		appNodeSuffix := fmt.Sprintf(`] %s.%s (expand)"`, resource.Type, resource.Name)
+		agentNodeSuffix := fmt.Sprintf(`] coder_agent.%s (expand)"`, agent.Name)
+
+		// Traverse the graph to check if the coder_<resource_type> depends on coder_agent.
+		for _, dst := range graph.Edges.SrcToDsts {
+			for _, edges := range dst {
+				for _, edge := range edges {
+					if strings.HasSuffix(edge.Src, appNodeSuffix) &&
+						strings.HasSuffix(edge.Dst, agentNodeSuffix) {
+						return true
+					}
+				}
+			}
+		}
+		return false
+	}
```

Comment:
My main concern is that we are traversing this graph multiple times in a tight loop.
Could we instead traverse the graph once, build a shallow map of `agent -> resource`, and use that to determine the agent-resource dependency?

Reply:
The main reason I'm hesitant to do this, it would be another cache/state used only to prevent this problem. Considering we have 1 agent on average, 2 in total, it may not introduce a heavy performance impact.

Anyway, if you think this is a must-do, I will improve the loop 👍  I admit, I focused mostly on fixing the problem.

Reply:
That's fair. Let's make it work correctly first, and then see about benchmarking / speeding it up.